wpilibc/src/main/native/cpp/LEDPattern.cpp

```diff
@@ -0,0 +1,290 @@
+    base.ApplyTo(data, writer);
+
+    // ... then, overwrite with illuminated LEDs from the overlay
+    ApplyTo(data, [&](int i, Color color) {
+      if (color.red > 0 || color.green > 0 || color.blue > 0) {
+        writer(i, color);
+      }
+    });
+  }};
+}
+
+LEDPattern LEDPattern::Blend(LEDPattern& other) {
+  return LEDPattern{[&](auto data, auto writer) {
+    // Apply the current pattern down as normal...
+    ApplyTo(data, writer);
+
+    other.ApplyTo(data, [&](int i, Color color) {
+      // ... then read the result and average it with the output from the other
+      // pattern
+      writer(i, Color{(data[i].r / 255.0 + color.red) / 2,
+                      (data[i].g / 255.0 + color.green) / 2,
+                      (data[i].b / 255.0 + color.blue) / 2});
+    });
+  }};
+}
+
+LEDPattern LEDPattern::Mask(LEDPattern& mask) {
+  return LEDPattern{[&](auto data, auto writer) {
+    // Apply the current pattern down as normal...
+    ApplyTo(data, writer);
+
+    mask.ApplyTo(data, [&](int i, Color color) {
+      auto currentColor = data[i];
+      // ... then perform a bitwise AND operation on each channel to apply the
+      // mask
+      writer(i, Color{currentColor.r & static_cast<uint8_t>(255 * color.red),
+                      currentColor.g & static_cast<uint8_t>(255 * color.green),
+                      currentColor.b & static_cast<uint8_t>(255 * color.blue)});
+    });
+  }};
+}
+
+LEDPattern LEDPattern::AtBrightness(double relativeBrightness) {
+  return LEDPattern{[=, this](auto data, auto writer) {
+    ApplyTo(data, [&](int i, Color color) {
+      writer(i, Color{color.red * relativeBrightness,
+                      color.green * relativeBrightness,
+                      color.blue * relativeBrightness});
+    });
+  }};
+}
+
+// Static constants and functions
+
+LEDPattern LEDPattern::kOff = LEDPattern::Solid(Color::kBlack);
+
+LEDPattern LEDPattern::Solid(const Color color) {
+  return LEDPattern{[=](auto data, auto writer) {
+    auto bufLen = data.size();
+    for (size_t i = 0; i < bufLen; i++) {
+      writer(i, color);
+    }
+  }};
+}
+
+LEDPattern LEDPattern::ProgressMaskLayer(
+    std::function<double()> progressFunction) {
+  return LEDPattern{[&](auto data, auto writer) {
+    double progress = std::clamp(progressFunction(), 0.0, 1.0);
+    auto bufLen = data.size();
+    auto max = (size_t)(bufLen * progress);
+
+    for (size_t led = 0; led < max; led++) {
+      writer(led, Color::kWhite);
+    }
+    for (size_t led = max; led < bufLen; led++) {
+      writer(led, Color::kBlack);
+    }
+  }};
+}
+
+LEDPattern LEDPattern::Steps(std::span<std::pair<double, Color>> steps) {
+  if (steps.size() == 0) {
+    // no colors specified
+    return LEDPattern::kOff;
+  }
+  if (steps.size() == 1 && steps[0].first == 0) {
+    // only one color specified, just show a static color
+    return LEDPattern::Solid(steps[0].second);
+  }
+
+  return LEDPattern{[=](auto data, auto writer) {
```

Comment:
This is capturing a span, which is a reference type which can go away.  You need to copy the span into a std::vector in the capture, e.g. `[steps = std::vector{steps.begin(), steps.end()}]`.

Reply:
Curly brace initializers didn't work out - the vectors contained pointers to the original pairs, instead of copies - but a normal constructor with parens seems to work.